feat: add performance analysis post (#13)
RafaelGSS/blog: priv/posts/performance-methodologies.md

---
title: Performance Methodologies
date: 2020-12-29 18:00
tags: performance,article,en-US
---

# Performance Methodologies

I've spent a good time studying _Performance_ methodologies and is fair to say that like everything on CS (Computer Science):

> > _Performance is a field where the more you know, the more you don't know_.

Author: Not sure about this suggestion. The meaning of the phase will change. Actually, the more you know, more you will understand, but there will have more paths that you don't know though.

On architecture diagram creation is obvious to say that _unknowns-unknowns_  is not mappable, after all, you don't know.

> Usually there is a lot of unknown-unknows and at least, is your job as a researcher in this field transform the `unknown-unknowns` into `know-unknows`.

The Diagram above map `known-knowns` (Green box) and `known-unknows` (Red box)

## Observability Tools

As previously mentioned, has observability in our software/architecture is fundamental to perform performance improvements. In this section, I'll briefly show you great tools to use in applications.

### Tracing

Tracing collects per-event data for analysis. Normally, Tracing tools are not enabled by default, since it adds CPU overhead to capture the data.

Logging (include system log) can be categorized as low-frequency tracing that is enabled by default.

Some common tools:

**system-wide**:
- `tcpdump`: network packet tracing
- `perf`: Linux Performance Events (tracing static and dynamic probes)

**per-process**:
- `strace`: system call tracing
- `USDT` (Userland Statically Defined Tracing)
- `DTrace`: observability framework that includes a programming language and a tool.

> 

Reviewer: I feel you're missing the "web" stack of tracing here. Jaeger, opentelemetry, zipkin, ..

For further information check the _useful links_ section.

### Profiling

Profiling characterizes the target by collecting a set of samples of snapshots. CPU usage is a common example where samples are taken of the stack trace to characterize the code paths that are consuming CPU cycles.

**Note**: For further information about Profiling CPU, I've made a blog post doing CPU Profiling in a Node.js application. [Check here](https://blog.rafaelgss.com.br/node-cpu-profiler).

Tools:

- `perf`: Linux Performance Events (profiling)
- `cachegrind`: a Valgrind sub tool, can profile hardware cache usage and be visualized using `kcachegrind`

> `/proc` is a file system interface for kernel statistics, it contains directories where each directory is named after the **PID** of the process. These directories contain a number of files containing information and statistics about each process mapped from kernel data structures.

![Julia Evans - Comic /proc](https://pbs.twimg.com/media/DZ3HpVXXkAEgxpc?format=jpg&name=large) - [reference](https://twitter.com/b0rk/status/981159808832286720/photo/1)

## Methodologies

> This section will describe three of most used methodologies. Apply a methodology due performance issues is needed, and there is no rule to choose the best approach,

Reviewer: ```suggestions
This section will describe three of the most used methodologies. Apply a methodology when performance issues start showing up; there is no rule about choosing the best approach.
````

> I would say that past experiences and knowledge in the software definitely will help with that.

Reviewer: ```suggestion
Previous experience with your software architecture will likely be the best way to make a decision.
```

### USE

Utilization, Saturation and Errors (USE) is an methodology that **should be used early in performance investigation**. For every resource, check the utilization, saturation, and errors:

> - **Resource**: all physical server functional components (CPU, busses, ...)

Reviewer: ```suggestion
- **Resource**: server components (CPU, buses, ...)
```

- **Utilization**: for a set time interval, the percentage of time that the resource was busy servicing work. While busy, the resource may **still be able to accept more work**.

> - **Saturation**: Extra work, often waiting on a queue. Are jobs that service cannot deal at moment.

Reviewer: ```suggestion
- **Saturation**: additional work to be done, likely waiting in a queue. Jobs that cannot be dealt with instantly.
```

![Workflow with USE Methodology](/images/performance-analysis/workflow-use.png)

> However, this could be counter-intuitive, a short burst of high utilization can cause saturation and performance issues, even though the overall utilization is low over a long interval. CPU utilization, for example, **can vary dramatically from second to second, so a 5-minute average may disguise short periods of 100% utilization and therefore saturation**.

Reviewer: ```suggestion
Its important to consider that it can be counter-intuitive; a short burst of high utilization can introduce saturation and performance issues even though the overall utilization is low over a long interval. CPU utilization **can change dramatically from second to second** so a 5-minute average may disguise short periods of 100% utilization and therefore lead to saturation.
```

Note: The saturation could not be easier to identify.

> The first step is to create a list of resources:

Reviewer: I would talk about how virtual resources are fundamentally different than dedicated hardware. Especially as your resources are both shared and intentionally throttled. Some - if not all - cloud providers make their money by overselling and betting on idle processes.

Author: Great comment! Added!

- **CPUs:** sockets, cores, hardware threads (virtual CPUs)

> - **Main memory**: DRAM

Reviewer: ```suggestion
- **Main memory**: RAM
```

- **Network interfaces**: Ethernet ports
- **Storage devices**: disks
- **Controllers**: storage, network
- **Interconnects**: CPU, memory, I/O

The USE method is most effective for resources that suffer performance degradation under high utilization or saturation, leading to bottlenecks. Fortunately, they are not common system bottlenecks, as they are typically designed to provide an excess of throughput. Unfortunately, if they are the problem, can be difficult to solve.

After you get the list of resources, try to create some metrics for it:

![List of resources to create metric](/images/performance-analysis/list-resources-use.png)

The process of elimination is good for us. Eliminate a possible resource bottleneck may help us to focus on another resource limiting our scope.

### Drill Down

> The process can involve digging down through deeper layers of the software stack, to hardware if necessary, to find the root cause of the issue. Particularly, I try to apply this methodology whenever that I can; It's hard to apply it without a big picture of your software, but when you get more experience on it, should be easier.

Reviewer: ```suggestion
The process iterates through deeper layers of the software stack – even to hardware if necessary – to find the root cause of the issue. I try to to apply this methodology in every part of the software stack. It's usually harder to do so  without having the bigger picture; but as you get more experienced you start recognizing recurring issues.
```

> > Analysis (Monitoring) is the base of all! Without it, we can't fix any bug.

Reviewer: ```suggestion
> Collecting and monitoring metrics is fundamental. Without it, we cannot fix the bugs the components cause.
```

> Such deeper analysis may involve the creation of custom tools and inspection of source code (if available). Here is where most of the drilling takes place, peeling away layers of the software stack as necessary to find the root cause.

Reviewer: ```suggestion
Deeper analysis may involve the creation of custom tools and introspection of source code (if available). Here is where most of the "digging" takes place, peeling away layers of the software stack as necessary to identify the root cause.
```

> For instance, let's imagine an application that after a month in production environment has begun to perform poorly.

Reviewer: ```suggestion
Imagine an application that after a month in an production environment has begun to perform poorly.
````

**Five Whys**

1. A database has performing poorly for some queries. Why?
2. It's delayed by disk I/O due to memory paging. Why?
3. Database memory usage has grown too large. Why?
4. The allocator is consuming more memory than expected. Why?
5. The allocator has a memory fragmentation issue.

This is a good real sample extracted from [System Performance](https://www.goodreads.com/book/show/18058001-systems-performance) book. There is not limit to go deep into _Why?_, but, one has to when the software is performing well.

![Five why - Drill Down](/images/performance-analysis/5-whys.png)

### Scientific Method

The _scientific method_ studies the _unknown_ by making hypotheses and testing them. The `unknown` here can mean the `unknown-unknown` as discussed in [`Know-Unknows`](#know-unknowns) section.

Every _scientific method_ consists:

1. Formulation of a question?
2. Hypothesis
3. Prediction
4. Testing
5. Analysis

> > To futher information about how _scientific methods_ works, see [here](https://en.wikipedia.org/wiki/Scientific_method)

Reviewer: ```suggestion
> For more information about how _scientific methods_, see [here](https://en.wikipedia.org/wiki/Scientific_method)
```

> First, you create a question based on performance problem, for instance: _Why my application have a degradation of throughput?_.

Reviewer: ```suggestion
First, define a question based on performance problem; for instance: _Why does my application have degraded throughput?_.
```

> After, you hypothesize what the cause of poor performance may be, _CPU Miss rate_. Then you construct a test to prove your theory, on this case I would recommend `Valgrind`.

Reviewer: ```suggestion
Second, build a hypothesis about what the cause of poor performance may be. _CPU Miss rate_? Write a test to prove your theory by for instance using `Valgrind`.
```

> The results collected by _Testing_ step is analyzed and then you have a better idea of what's happening.

Reviewer: ```suggestion
Collect results from your previous step and analyze how it behaves over time. It will give you a better idea of what components are connected and ultimately affected.
```

> **Note:** Create a hypothesis is not lucky, you should have a clear understanding of your software architecture. A versioning of architectural changes is as important as.

Reviewer: ```suggestion
**Note:** Shaping a hypothesis requires a clear understanding of your software architecture. Versioning your architectural changes can play a key role in understanding sudden changes.
```

---
title: Performance Methodologies
date: 2021-01-21 10:00
tags: performance,article,en-US
---

# Performance Methodologies

> Poor performance costs the software industry millions of dollars annually in lost revenue, decreased productivity, increased development and hardware costs, and damaged customer
> relations.

Reviewer: ```suggestion
Poor performance costs the software industry millions of dollars annually in lost revenue, decreased productivity, increased development and hardware costs and damaged customer relations.
```

> Most applications tend to focus on correctness over performance. The shift towards performance only occurs once it becomes a problem.

Reviewer: ```suggestion
Most applications tend to focus on correctness over performance. The shift towards performance only occurs once it is seen as a problem.
```

> Once it becomes problem, one rarely has time to dedicate towards improving it. This article aims to show you that **there is no magic solution**.

Reviewer: ```suggestion
When that happens, one rarely has time to dedicated towards improving it. This article aims to show you that **there is no simple answer**.
```

> A lot of work should be done in early phases of development. This article will consider the reader having the role of  Performance Engineer or having to "act" as one.

Reviewer: ```suggestion
A lot of performance work should be done in early phases of development. For the rest of the article, the reader is considered having a role of Performance Engineer or "acting" as one.
```

> The reality is that some companies devotees of "agile" methodology, adopt the same order to every software development process: Idea -> MVP -> Feature -> "Refactor". One only needs some time in this methodology to understand that it's not quite like that.
> The result is that, somewhere near the end of the project, performance issues appear.

Reviewer: ```suggestion
In my experience, a strictly "agile" methodology (Idea -> MVP -> Feature -> "Refactor") tends to leave out proper performance engineering, since performance is not a goal but an expectation.
```

> _Performance is a field where the more you know, the less you understand_.

> Regardless of the source, when a performance issue appears it should be fixed immediately and there are few ways: fix the problem in the source (or at least understand it) or spending money on hardware resources. The second path in some time will lead to the same problem.

Reviewer: ```suggestion
Regardless of the source, when a performance issue appears it should be fixed immediately. Two ways are: 1) modifying code/architecture or throwing money at additional hardware resources. The second path in some time will lead to the same problem down the line.
```

> Prior starting performance analysis, you **must**  understand your application architecture. Any analysis requires clear boundaries and a full understanding of dependencies and third-party services.

Reviewer: ```suggestion
Prior starting performance analysis, you **must** understand your application architecture. Any analysis requires clear boundaries and a full understanding of dependencies and third-party services.
```

A diagram of your software architecture is a great starting point.

> The foundation of your software should be resilient to achieve better results.

## Monitoring

Today, a big part of the market is adopting distributed systems. As we've come to learn, such systems adds a lot of complexity to your architecture in exchange for scalability and availability (resilience).

It also adds more components to your list of dependencies. Therefore, you should monitor these dependencies to have better visibility when things deviate from a happy path.

Each part of the architecture (or software) needs individual monitoring that helps us go back in time and answer some of these questions:

> - When started the software performing worse?

Reviewer: ```suggestion
- When did the software start performing worse?
```

> - During what timing window are we seeing most access/traffic?

Reviewer: ```suggestion
- During what timeframe are we seeing most activity?
```

> - How our devices are working on a specific date? Like I/O latency, DNS resolution

Reviewer: ```suggestion
- How are components behaving during these timeframes? Think: I/O latency, DNS resolution, CPU or RAM consumption
```

## CPU

The next figure shows an example CPU architecture, for a single processor with four cores and eight hardware threads in total.

![CPU Architecture](/images/performance-analysis/cpu-architecture.png)

Each hardware thread is addressable as a logical CPU, so this processor appears as eight CPUs. The operating system may have some additional knowledge of topology, such as which CPUs are on the same core, to improve its scheduling decisions.

Processors provide various hardware caches for improving memory I/O performance.

![CPU Layers](/images/performance-analysis/cpu-cache-layers.png)

> High CPU utilization may not necessarily be a problem, but rather a sign that the system is doing work.

---

**MMU (Memory Management Unit)**

The MMU is responsible for virtual-to-physical address translation.

This MMU uses an on-chip TLB to cache address translations. Cache misses are satisfied by translation tables in main memory (DRAM), called page tables, which are read directly by the MMU (hardware).

![Memory Architecture](/images/performance-analysis/mmu-architecture.png)

---

> **Tools to CPU analysis**

Reviewer: ```suggestion
**Tools for CPU analysis**
```

For CPUs, the tools method can involve checking the following: 

- uptime: Check load averages to see if CPU load is increasing or decreasing over time. A load average over the number of CPUs in the system usually indicates saturation.
- vmstat: Run vmstat per second, and check the idle column to see how much headroom there is. Less than 10% can be a problem.
- mpstat: Check for individual hot (busy) CPUs, identifying a possible thread scalability problem.
- top/htop: See which processes and users are the top CPU consumers.
- perf: Measure CPI

## Performance Issue Example

> The last section, CPU Cache was mentioned. Take a look in the following example:

Reviewer: ```suggestion
In the last section, we brought up CPU Cache. Consider the following example:
```

> > _"An application are performing bad, and based in the monitoring report, is performing a CPU Intensive workload_"

Reviewer: ```suggestion
> _"An application is performing bad. Monitoring is telling us that has a CPU Intensive workload_"
```

double C[n][n];

float tdiff(struct timeval* start, struct timeval* end) {
  return (end->tv_sec - start->tv_sec) +
    1e-6 * (end->tv_usec - start->tv_usec);
}

int main(int argc, const char* argv[]) {
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      A[i][j] = (double)rand() / (double)RAND_MAX;
      B[i][j] = (double)rand() / (double)RAND_MAX;
      C[i][j] = 0;
    }
  }

  struct timeval start, end;

  gettimeofday(&start, NULL);

  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      for (int k = 0; k < n; ++k) {
        C[i][j] += A[i][k] * B[k][j];
      }
    }
  }

  gettimeofday(&end, NULL);
  printf("%0.6f\n", tdiff(&start, &end));
  return 0;
}
```

The code are spending around of 50 seconds in the server machine. It should be improved by a factor of 10 (following the SLO of application).

> The time elapsed varies by machine, of course.

To this performance issue, I would perform the _Scientific Method_ approach:

1. Question. Why this application are performing poorly?
2. Hypothesis. The ticket shows that application are spending time in CPU intensive operations. The source code tell us that it could has high CPU Cache miss.

> 3. Prediction. Improve the spatial locality through vector access will get better cache hit.

Reviewer: ```suggestion
3. Prediction. Improve the spatial locality through vector access will yield a higher cache hit ratio.
```

sudo perf stat -e cache-references,cache-misses ./main.o

52.590355

 Performance counter stats for './main.o':

    11.641.838.626      cache-references
     1.036.364.121      cache-misses              #    8,902 % of all cache refs

      52,674319792 seconds time elapsed

      51,162040000 seconds user
       0,083826000 seconds sys
```

10% of cache misses is a great point of improvement. This code is performing a bad spatial locality.

**The order of for loop matters!** Switch the `for-j` to `for-k` will improve the cache hit.

```diff
  for (int i = 0; i < n; ++i) {
-   for (int j = 0; j < n; ++j) {
+   for (int k = 0; k < n; ++k) {
-     for (int k = 0; k < n; ++k) {
+     for (int j = 0; j < n; ++j) {
        C[i][j] += A[i][k] * B[k][j];
      }
    }
  }
```

Running the application again, it gets:

```sh
sudo perf stat -e cache-references,cache-misses ./main.o

3.184440

 Performance counter stats for './main.o':

     2.572.702.714      cache-references
        46.695.681      cache-misses              #    1,815 % of all cache refs

       3,274432507 seconds time elapsed

       3,157639000 seconds user
       0,040020000 seconds sys
```

Almost **17x faster**!

> Of course, this is a sample. In real-world issues, there are several variables to consider and is normal when a scientific method doesn't give an immediate result, but only through attempts, it gives better visibility and improves the application.

Reviewer: ```suggestion
This is obviously for demonstration purposes only. In real-world issues there are several variables to consider -- it is normal when a scientific method doesn't give an immediate result. Only through different attempts you will increase visibility and improve the application.
```
